### ISSUE TITLE
With the latest changes sometimes a listener `start_keyword` `source` is an empty string (when it shouldn't be).

### ISSUE THREAD

**Issue by fabioz** [robotframework/robotframework#4599]
Note: the code that introduces this bug is:

https://github.com/robotframework/robotframework/commit/92148c8566a3f39ca098e0da38b0e1eac5619d1a (if I revert to a version before it my tests work).

I'm posting below the image on the debugger showing the empty source:

![image](https://user-images.githubusercontent.com/117621/212469051-8e9d0887-3ebe-4f41-bea3-b24ca89c36a6.png)

This happens because for some reason the `Keyword.parent` is `None` (see screenshot below) -- and now the `source` requires the `parent`...

![image](https://user-images.githubusercontent.com/117621/212469298-97b5c214-4a6f-4f4a-b8bd-0c63cafee190.png)

**Comment by fabioz** [comment]
I think changing the `source` to a `Path` instead of a `string` can break many things silently, so, it does make sense that this is the problem.

As a note, I had other tests break here due to that change too... this should definitely require a new major release just for this change... personally, I'd just create a `source_path` and leave compatibility for the old `source` instead of changing its type (due to the breakage of clients I'm not even sure the backward incompatibility change is worth it).

**Comment by pekkaklarck** [comment]
What the source was earlier? A proper path (as a string) or `None`? Does it happen with all keywords or only with some?

We can discuss is making `source` a `pathlib.Path` instance in in RF 6.1 a good idea in #4596. I marked that issue `backwards incompatible`, but considered it safe enough in RF 6.1 because it works as earlier when used as a path e.g. with `open()` or `os.path` functions. Using it as a string obviously won't work, but `str(suite.source)` is an easy fix. Anyway, it can be considered too big change in a feature release and be postponed to RF 7.0.

**Comment by fabioz** [comment]
It was a proper path -- my use case just starts the code from one file and uses the listener (this is what my test case expected and it started to fail because the path is now empty).

I don't know why the change makes it not have a parent now (given that the parent change predates the change from the change to being a Path).

**Comment by fabioz** [comment]
This could be potentially be because of: https://github.com/robotframework/robotframework/issues/4600... when paths resolve they are no longer the same path initially requested and that can have bad repercussions -- I use a subst when running locally just to get those kind of errors before clients complain ;)

Anyways, I don't really know the real culprit, I just know that that change made it so that the listener is now receiving an empty source when it used to receive the actual path before.

**Comment by fabioz** [comment]
I'm attaching a project where the error can be reproduced (it doesn't seem related to https://github.com/robotframework/robotframework/issues/4600 because I can reproduce it even when running without the subst).

To reproduce unzip the project and run `python my_entry_point.py`.

It seems related to the `Run Keyword If`.

[repro_project.zip](https://github.com/robotframework/robotframework/files/10417307/repro_project.zip)

**Comment by pekkaklarck** [comment]
I looked at our tests and they ought to validate that `start_keyword` gets course correctly. I haven't yet had time to look at the above example project, but `Run Keyword If` being somehow related sounds plausible.

There have been some fixes lately. If you @fabioz have time, could you check does the problem still occur in your environment?

**Comment by fabioz** [comment]
I just checked it. The error persists. Have you tried to run the sample project locally (the listener throws and error and it appears in the output in the test).

**Comment by pekkaklarck** [comment]
Hadn't had time to look at the sample project earlier, but I looked at it now. I can reproduce the issue, but for me it occurs also with earlier Robot versions. Are you sure it works in your environment with RF 6.0.2 or earlier?

As you suggested, the issue seems to be related to `Run Keyword If` because with other keywords and control structures the source is available. I'll look could it be fixed easily.

**Comment by pekkaklarck** [comment]
Looking at the code it seems this couldn't have worked earlier either. Good news is that making it work doesn't seem too complicated.

**Comment by pekkaklarck** [comment]
I fixed the problem with `Run Keyword` and its variants. There now ought to always be valid `source`, unless users are doing something _really_ funky, but it's possible that `lineno` is `None`. The only way I think that can happen is that you have a listener that calls `BuiltIn.run_keyword` in `start_suite`. Not sure should we use `1` instead of `None` in that case.

**Comment by fabioz** [comment]
Great, thanks for taking a look at it.

As a note, after this comment:

> Looking at the code it seems this couldn't have worked earlier either. Good news is that making it work doesn't seem too complicated.

I started to take a deeper look into my code to know why it broke then. 

The issue was actually that the behavior changed, not that the source was reported before.

i.e.: Prior to the path change, when the source wasn't available, `None` would be reported (which was properly handled in the debugger and it went up and got up the stack), but with that change it started returning an empty string and that wasn't handled properly, so, it didn't go up the stack to reuse the same filename.

-- I believe that having that in core with your current fix does make more sense than the workaround I had in place for this situation though ;)

I think using `None` is fine if the line really isn't available (the docs on `lineno` and `source` should probably be updated to reflect the current situation though as it's not clear that sometimes the line or source won't be available from the docs).

**Comment by pekkaklarck** [comment]
`None` being changed to `''` definitely sounds possible. No need to change that after #4604 being fixed. I believe this issue can be closed.

`lineno` is `None` if we don't find any other parent than the root test suite. The only way I can see that happening is some listener running keywords. `source` could be `None` if someone creates a suite programmatically and doesn't set `source`. These are so special cases that I'm not sure are they worth documenting separately.
